<a href="https://colab.research.google.com/github/danchaud-vincent/tensorflow-deep-learning/blob/main/exercices-solutions/04_Transfer_Learning_with_TensorFlow_Part_1_Feature_Extraction_Exercice_Solutions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 04 - Transfer Learning with TensorFlow Part 1: Feature Extraction Exercice Solutions

1. Build and fit a model using the same data we have here but with the MobileNetV2 architecture feature extraction from TensorFlow Hub. How does it perform compared to our other mdoels?
2. Name 3 different image classification models on TensorFlow Hub that we haven't used.
3. Build a model to classify images of two different things you've taken photos of.
- You can use any feature extraction layer from TensorFlow Hub you like for this.
- You should aim to have at least 10 images of each class, for example to build a fridge versus oven classifier, you'll want 10 images of fridges and 10 images of ovens.
4. What is the current best performing model on ImageNet?

# 1. Build and fit a model using the same data `10_food_classes` but with the MobileNetV2 architecture feature extraction from TensorFlow Hub


## 1.1 Get the data

In [2]:
# download the data
!wget https://storage.googleapis.com/ztm_tf_course/food_vision/10_food_classes_10_percent.zip

--2022-08-30 18:28:49--  https://storage.googleapis.com/ztm_tf_course/food_vision/10_food_classes_10_percent.zip
Resolving storage.googleapis.com (storage.googleapis.com)... 173.194.210.128, 173.194.211.128, 173.194.212.128, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|173.194.210.128|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 168546183 (161M) [application/zip]
Saving to: ‘10_food_classes_10_percent.zip’

10_food_classes_10_ 100%[===================>] 160.74M   144MB/s    in 1.1s    

2022-08-30 18:28:50 (144 MB/s) - ‘10_food_classes_10_percent.zip’ saved [168546183/168546183]



In [6]:
# extract the data inside the zip file
import zipfile

zip_ref = zipfile.ZipFile("10_food_classes_10_percent.zip")
zip_ref.extractall()
zip_ref.close()

In [8]:
# explore the data folder
import os

for dirpath, dirnames, filenames in os.walk("10_food_classes_10_percent"):
  print(f"{dirpath} contains : {len(dirnames)} directories, {len(filenames)} files")

10_food_classes_10_percent contains : 2 directories, 0 files
10_food_classes_10_percent/test contains : 10 directories, 0 files
10_food_classes_10_percent/test/ice_cream contains : 0 directories, 250 files
10_food_classes_10_percent/test/hamburger contains : 0 directories, 250 files
10_food_classes_10_percent/test/chicken_wings contains : 0 directories, 250 files
10_food_classes_10_percent/test/pizza contains : 0 directories, 250 files
10_food_classes_10_percent/test/ramen contains : 0 directories, 250 files
10_food_classes_10_percent/test/steak contains : 0 directories, 250 files
10_food_classes_10_percent/test/chicken_curry contains : 0 directories, 250 files
10_food_classes_10_percent/test/grilled_salmon contains : 0 directories, 250 files
10_food_classes_10_percent/test/sushi contains : 0 directories, 250 files
10_food_classes_10_percent/test/fried_rice contains : 0 directories, 250 files
10_food_classes_10_percent/train contains : 10 directories, 0 files
10_food_classes_10_percent

Our training directories have 75 images per classes rather than the test directories which have 250 images per classes.

In [10]:
# Get the class names
import pathlib 

data_dir = pathlib.Path("10_food_classes_10_percent/train/")

for i in data_dir.glob("*"):
  print(i.name)

ice_cream
hamburger
chicken_wings
pizza
ramen
steak
chicken_curry
grilled_salmon
sushi
fried_rice


In [12]:
# get the class names in an array
import numpy as np
class_names = np.array(sorted([i.name for i in data_dir.glob("*")]))
class_names

array(['chicken_curry', 'chicken_wings', 'fried_rice', 'grilled_salmon',
       'hamburger', 'ice_cream', 'pizza', 'ramen', 'steak', 'sushi'],
      dtype='<U14')

## 1.2 Preparing the data